File: python/mirage/core/nextcloud/truncate.py
from opendal.exceptions import NotFound

from mirage.accessor.nextcloud import NextcloudAccessor
from mirage.cache.context import invalidate_after_write
from mirage.observe.context import record, start_op
from mirage.types import PathSpec


async def truncate(accessor: NextcloudAccessor, path: PathSpec,
                   length: int) -> None:
    key = path.mount_path.lstrip("/")
    timer = start_op()
    op = accessor.operator()
    try:
        data = bytes(await op.read(key))
    except NotFound:
        data = b""
    result = data[:length].ljust(length, b"\0")
    await op.write(key, result)
    record("truncate", path.virtual, "nextcloud", 0, timer)
    await invalidate_after_write(path)

File: typescript/packages/node/src/core/nextcloud/truncate.ts
import { invalidateAfterWrite } from '@struktoai/mirage-core/cache/context'
import { record, startOp } from '@struktoai/mirage-core/observe/context'
import { ResourceName } from '@struktoai/mirage-core/types'
import type { PathSpec } from '@struktoai/mirage-core/types'
import type { NextcloudAccessor } from '../../accessor/nextcloud.ts'
import { isNotFound, nextcloudKey } from './util.ts'

export async function truncate(
  accessor: NextcloudAccessor,
  path: PathSpec,
  length: number,
): Promise<void> {
  const timer = startOp()
  const op = await accessor.operator()
  const key = nextcloudKey(path)
  let current: Buffer
  try {
    current = await op.read(key)
  } catch (error) {
    if (!isNotFound(error)) throw error
    current = Buffer.alloc(0)
  }
  const next = Buffer.alloc(length)
  current.copy(next, 0, 0, Math.min(current.byteLength, length))
  await op.write(key, next)
  record('truncate', path.virtual, ResourceName.NEXTCLOUD, 0, timer)
  await invalidateAfterWrite(path)
}
